pandas/io/common.py

```diff
@@ -878,6 +887,46 @@ def __next__(self) -> str:
         return newline.lstrip("\n")
 
 
+# Wrapper that wraps a StringIO buffer and reads bytes from it
+# Created for compat with pyarrow read_csv
+class BytesIOWrapper(io.BytesIO):
+    buffer: StringIO | TextIOBase | None
+
+    def __init__(self, buffer: StringIO | TextIOBase, encoding: str = "utf-8"):
+        self.buffer = buffer
+        self.encoding = encoding
+        # Because a character can be represented by more than 1 byte,
+        # it is possible that reading will produce more bytes than n
+        # We store the extra bytes in this overflow variable, and append the
+        # overflow to the front of the bytestring the next time reading is performed
+        self.overflow = b""
+
+    def __getattr__(self, attr: str):
+        return getattr(self.buffer, attr)
+
+    def read(self, n: int | None = -1) -> bytes:
+        assert self.buffer is not None
+        bytestring = self.buffer.read(n).encode(self.encoding)
+        # When n=-1/n greater than remaining bytes: Read entire file/rest of file
+        combined_bytestring = self.overflow + bytestring
+        if n is None or n < 0 or n >= len(combined_bytestring):
+            self.overflow = b""
+            return combined_bytestring
+        else:
+            to_return = combined_bytestring[:n]
+            self.overflow = combined_bytestring[n:]
+            return to_return
+
+    def detach(self):
```

Comment:
Why can't you use `self.buffer`'s detach?

Reply:
That would close/mess up the buffer passed in by the user, which we won't want to touch. I think detach exists for GC purposes, but not sure.

Reply:
I think this block:
https://github.com/pandas-dev/pandas/pull/42669/files#diff-934d8564d648e7521db673c6399dcac98e45adfd5230ba47d3aabfcc21979febR105
should take care of not closing user-provided file handles?

Reply:
That calls detach, which is not implemented by BytesIO(the parent class of BytesIOWrapper).

Reply:
okay, sorry I wasn't aware of that.